mgmt: hawkbit: domain name buffer length and add check

nealjack opened the pull request:
This PR fixes the `hb_cfg.server_addr` buffer and adds a length check when setting this value.

Hawkbit uses `DNS_MAX_NAME_SIZE` for its internal buffer which is intended for a DNS response, not the query. By default `DNS_MAX_NAME_SIZE` is only 20 bytes and results in truncation for a domain name longer than that. For example, the domain name hawkbit.yourcompany.com would be truncated and fail to resolve. The Kconfig `DNS_RESOLVER_MAX_QUERY_LEN` is the correct macro to use here for a query, and is by default 256 bytes, in line with [RFC1035](https://www.ietf.org/rfc/rfc1035.txt).

Since `DNS_RESOLVER_MAX_QUERY_LEN` is adjustable, `hawkbit_set_config` also requires a length check to ensure that any domain name that is passed in can be fit in its internal buffer. This prevents silent truncation of the domain name.

nealjack committed:
mgmt: hawkbit: replace DNS_MAX_NAME_SIZE with DNS_RESOLVER_MAX_QUERY_LEN

The hawkbit subsystem was erroneously using DNS_MAX_NAME_SIZE as the
maximum query length. This limited query strings to only 20 bytes,
truncating many domain names.

Signed-off-by: Neal Jackson <neal@blueirislabs.com>

nealjack committed:
mgmt: hawkbit: add check for valid domain name length

This commit adds a `strnlen` length check for `server_addr` to ensure that
it will not be truncated and result in a silent failure. Instead, the
call to `hawkbit_set_config` will return -EINVAL if the supplied
`server_addr` is too long for the internal buffer.

Signed-off-by: Neal Jackson <neal@blueirislabs.com>

maass-hamburg commented on subsys/mgmt/hawkbit/hawkbit.c:
```diff
@@ -78,7 +78,7 @@ static bool hawkbit_initialized;
 	int32_t action_id;
 #ifdef CONFIG_HAWKBIT_SET_SETTINGS_RUNTIME
-	char server_addr[DNS_MAX_NAME_SIZE + 1];
+	char server_addr[CONFIG_DNS_RESOLVER_MAX_QUERY_LEN + 1];
```
hawkBit does not require the DNS_RESOLVER, so `CONFIG_DNS_RESOLVER_MAX_QUERY_LEN` could be undefined

maass-hamburg replied:
note this https://github.com/zephyrproject-rtos/zephyr/pull/89533

nealjack replied:
What do you suggest be used instead of this flag? Do you want me to hardcode 255?

maass-hamburg replied:
> If CONFIG_DNS_RESOLVER_MAX_QUERY_LEN is defined we could use that, otherwise use INET6_ADDRSTRLEN ?

this seems like a good idea

maass-hamburg replied:
`#include <zephyr/net/dns_resolve.h>` can than also be removed, DNS_MAX_NAME_SIZE was the only thing we used from that include.

maass-hamburg reviewed (changes requested):
also after line 287 in hawkbit.c
so that the length is not checkt that strict of the server_addr:
line 288: ``if (len > sizeof(hb_cfg.server_addr)) {``
line 292: ``rc = read_cb(cb_arg, &hb_cfg.server_addr, len);``

line 361: ``(void)cb("hawkbit/server_addr", &hb_cfg.server_addr, strlen(hb_cfg.server_addr));``

nealjack committed:
mgmt: hawkbit: use sizeof in server_addr length checks

Simplify length checks by using sizeof for the server_addr array.

Signed-off-by: Neal Jackson <neal@blueirislabs.com>

maass-hamburg commented on include/zephyr/mgmt/hawkbit/config.h:
```diff
@@ -45,6 +45,7 @@ struct hawkbit_runtime_config {
  * @param config Configuration settings to set.
  * @retval 0 on success.
  * @retval -EAGAIN if probe is currently running.
+ * @retval -EINVAL if config->server_addr is too long.
```
should be a more generic description, as we could in the future also check config->auth_token f.e.

nealjack replied:
I changed it to be more generic in most recent force push

nealjack committed:
mgmt: hawkbit: remove DNS_RESOLVER requirement for server_addr

This commit removes the requirement for DNS_RESOLVER. If DNS_RESOLVER is
enabled, hawkbit uses `CONFIG_DNS_RESOLVER_MAX_QUERY_LEN` for the
server_addr buffer, if disabled it uses `CONFIG_INET6_ADDRSTRLEN`.
This adheres to #89533 which removes the requirement for DNS_RESOLVER.

Signed-off-by: Neal Jackson <neal@blueirislabs.com>

maass-hamburg commented on subsys/mgmt/hawkbit/hawkbit.c:
```diff
@@ -778,8 +784,16 @@ int hawkbit_default_config_data_cb(const char *device_id, uint8_t *buffer, const
 #ifdef CONFIG_HAWKBIT_SET_SETTINGS_RUNTIME
 int hawkbit_set_config(struct hawkbit_runtime_config *config)
 {
+	size_t length;
+
 	if (k_sem_take(&probe_sem, HAWKBIT_SET_SERVER_TIMEOUT) == 0) {
 		if (config->server_addr != NULL) {
+			length = strnlen(config->server_addr, sizeof(hb_cfg.server_addr));
+			if (length > sizeof(hb_cfg.server_addr) - 1) {
```
```suggestion
	if (k_sem_take(&probe_sem, HAWKBIT_SET_SERVER_TIMEOUT) == 0) {
		if (config->server_addr != NULL) {
			if (strnlen(config->server_addr, sizeof(hb_cfg.server_addr)) >= sizeof(hb_cfg.server_addr)) {
```

The ``>=`` could probably also be a ``==`` as the maximum val of the `strnlen()` would be `sizeof(hb_cfg.server_addr)`

nealjack replied:
I think you'd still want to make sure that there is a null byte at the end of the buffer. So strnlen should return sizeof(hb_cfg.server_addr) - 1 at the most.

maass-hamburg commented on subsys/mgmt/hawkbit/hawkbit.c:
```diff
@@ -285,11 +290,11 @@ static int hawkbit_settings_set(const char *name, size_t len, settings_read_cb r
 #ifdef CONFIG_HAWKBIT_SET_SETTINGS_RUNTIME
 	if (settings_name_steq(name, "server_addr", &next) && !next) {
-		if (len != sizeof(hb_cfg.server_addr)) {
+		if (len > sizeof(hb_cfg.server_addr)) {
 			return -EINVAL;
 		}
 
-		rc = read_cb(cb_arg, &hb_cfg.server_addr, sizeof(hb_cfg.server_addr));
+		rc = read_cb(cb_arg, &hb_cfg.server_addr, len);
```
what do you think about removing the check at the beginning and using `MIN(len, sizeof(hb_cfg.server_addr))` as len in `read_cb()` and using `strnlen()` after it to check if the string was NULL terminated?

maass-hamburg replied:
this way we could use INET_ADDRSTRLEN instead of INET6_ADDRSTRLEN, when IPv6 is not activated

nealjack replied:
Wouldn't you want to ensure the string is null terminated before calling `read_cb`?

maass-hamburg replied:
If it would bepossible yes, but only in read_cb the string will be read, we don't have it before

nealjack replied:
ah okay I was unfamiliar with what read_cb was doing. If it is copying a string into the buffer then sure - the check should be after.

maass-hamburg commented on subsys/mgmt/hawkbit/hawkbit.c:
```diff
@@ -358,7 +363,8 @@ static int hawkbit_settings_export(int (*cb)(const char *name, const void *value
 #ifdef CONFIG_HAWKBIT_SET_SETTINGS_RUNTIME
-	(void)cb("hawkbit/server_addr", &hb_cfg.server_addr, sizeof(hb_cfg.server_addr));
+	(void)cb("hawkbit/server_addr", &hb_cfg.server_addr,
+		 strnlen(hb_cfg.server_addr, sizeof(hb_cfg.server_addr)));
```
```suggestion
		 strnlen(hb_cfg.server_addr, sizeof(hb_cfg.server_addr)) + 1);
```

nealjack replied:
Why the +1? I don't think you want to pass a length off the end of the array. The length should be at maximum `sizeof(hb_cfg.server_addr))` if the string is not null terminated.

maass-hamburg replied:
We want to pass the length of the string including zero termination. Maybe we should use the normal strlen here.

nealjack replied:
the string, including the null termination byte, should be within the bounds of the array. I don't understand why you would want strnlen to read off the bounds of `hb_cfg.server_addr`.

maass-hamburg replied:
The + 1 is after the strnlen() not inside.

nealjack replied:
> The + 1 is after the strnlen() not inside.

Okay, but what happens when strnlen returns sizeof(hb_cfg.server_addr)? The cb will be passed `sizeof(hb_cfg.server_addr) + 1`. I'm not very familiar with the settings subsystem, so I'm not sure what is passed as `cb`.

maass-hamburg replied:
It shouldn't, when setting it we already check, that it's not too long. strlen should be fine enough. In the zephyr codebase strnlen is rarely used in difference to strlen.

maass-hamburg commented on subsys/mgmt/hawkbit/hawkbit.c:
```diff
@@ -285,11 +290,11 @@ static int hawkbit_settings_set(const char *name, size_t len, settings_read_cb r
+		rc = read_cb(cb_arg, &hb_cfg.server_addr, MIN(len, sizeof(hb_cfg.server_addr)));
+		if (strnlen(hb_cfg.server_addr, sizeof(hb_cfg.server_addr)) ==
+		    sizeof(hb_cfg.server_addr)) {
 			return -EINVAL;
```
Should clean hb_cfg.server_addr here by doing a memset 0 there.

maass-hamburg commented on subsys/mgmt/hawkbit/hawkbit.c:
```diff
@@ -75,10 +74,16 @@ static bool hawkbit_initialized;
+#ifdef CONFIG_DNS_RESOLVER_MAX_QUERY_LEN
+#define SERVER_ADDR_LEN CONFIG_DNS_RESOLVER_MAX_QUERY_LEN
+#else
+#define SERVER_ADDR_LEN INET6_ADDRSTRLEN
```
If ipv6 is not activated, it should only be the size needed for a ipv4 address

maass-hamburg commented on subsys/mgmt/hawkbit/hawkbit.c:
```diff
@@ -75,10 +74,19 @@ static bool hawkbit_initialized;
 
+#ifdef CONFIG_DNS_RESOLVER_MAX_QUERY_LEN
+#define SERVER_ADDR_LEN CONFIG_DNS_RESOLVER_MAX_QUERY_LEN
+#else
+#ifdef CONFIG_NET_IPV6
```
```suggestion
#elif defined(CONFIG_NET_IPV6)
```

maass-hamburg commented on subsys/mgmt/hawkbit/hawkbit.c:
```diff
@@ -285,11 +293,12 @@ static int hawkbit_settings_set(const char *name, size_t len, settings_read_cb r
 #ifdef CONFIG_HAWKBIT_SET_SETTINGS_RUNTIME
 	if (settings_name_steq(name, "server_addr", &next) && !next) {
-		if (len != sizeof(hb_cfg.server_addr)) {
+		memset(hb_cfg.server_addr, 0, sizeof(hb_cfg.server_addr));
+		rc = read_cb(cb_arg, &hb_cfg.server_addr, MIN(len, sizeof(hb_cfg.server_addr)));
+		if (strnlen(hb_cfg.server_addr, sizeof(hb_cfg.server_addr)) ==
+		    sizeof(hb_cfg.server_addr)) {
 			return -EINVAL;
 		}
```
```suggestion
		rc = read_cb(cb_arg, &hb_cfg.server_addr, MIN(len, sizeof(hb_cfg.server_addr)));
		if (strnlen(hb_cfg.server_addr, sizeof(hb_cfg.server_addr)) ==
		    sizeof(hb_cfg.server_addr)) {
			memset(hb_cfg.server_addr, 0, sizeof(hb_cfg.server_addr));
 			return -EINVAL;
 		}
```

nealjack commented:
Are there any other changes that are required for this pull request?
